File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/array/P5_LongestPalindromicSubstring.java
package com.dsa.leetcode.labuladong.array;

/**
 * @since 2023/11/5 08:57
 */
public class P5_LongestPalindromicSubstring {

    public String longestPalindrome(String s) {
        if (s == null || s.length() == 0) {
            return "";
        }
        int start = 0, end = 0;
        for (int i = 0, length = s.length(); i < length; i++) {
            int len1 = expandAroundCenter(s, length, i, i);
            int len2 = expandAroundCenter(s, length, i, i + 1);
            int len = Math.max(len1, len2);
            if (len > end - start) {
                start = i - (len - 1) / 2;
                end = i + len / 2;
            }
        }
        return s.substring(start, end + 1);
    }

    public int expandAroundCenter(String str, int length, int left, int right) {
        while (left >= 0 && right < length && str.charAt(left) == str.charAt(right)) {
            left--;
            right++;
        }
        return right - left - 1;
    }

}

File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/array/two_pointers/P5_LongestPalindromicSubstring.py
class Solution:
    def expandAroundCenter(self, s, length, left, right):
        while left >= 0 and right < length and s[left] == s[right]:
            left -= 1
            right += 1
        return left + 1, right - 1

    def longestPalindrome(self, s: str) -> str:
        if not s:
            return ""
        start, end = 0, 0
        length = len(s)
        for i in range(length):
            left1, right1 = self.expandAroundCenter(s, length, i, i)
            left2, right2 = self.expandAroundCenter(s, length, i, i + 1)
            if right1 - left1 > end - start:
                start, end = left1, right1
            if right2 - left2 > end - start:
                start, end = left2, right2

        return s[start: end + 1]
